### PR TITLE
Espresso test - Add mock location service

### PR DESCRIPTION
**Pull request by kshamashuttl**
[Clubhouse ticket](https://app.clubhouse.io/shuttl/story/7637/ca-build-tests-for-new-gps-sdk)

**PR details** : 

**New Files added :** 
 - MockLocationHelper : To start and stop mock location service
 - MockLocationProvider : To set mock location 
 - MockLocationProviderManager : To set lat-lng in sharedPreference
 - MockLocationSaveService : Separate service for mock location

**AndroidManifest changes** :
- Add **ACCESS_MOCK_LOCATION** permission 
- Register mock location service

Other commit in this PR : 
- [Add helper method](https://github.com/betaCode95/LocationsSync/commit/6399a2fc25ada25b3d96132ef63ac1937d26b5c1)

**Successful logs :** 
MockLocation: startMockLocationService : Start mock location service
MockLocation: onCreate : Start foreground service to sync mock location
MockLocation: subscribeMockLocationTracking : Request for location updates
MockLocation: addTestProviderToMockAvailableProviders : Add test provider = gps
MockLocation: setMockProviderLocationData : Mock location latLng from SharedPreferences Location[gps 29.770000,77.279999 hAcc=20 et=+1h33m14s483ms alt=634.61 bear=0.0 vAcc=??? sAcc=??? bAcc=???]
MockLocation: addMockLocationProvider : Add provider and set mock location latLng
MockLocation: work : Start mock location provider 
MockLocation: onLocationChanged : Update changed location in DB Location[gps 29.770000,77.279999 hAcc=20 et=+1h33m14s483ms alt=634.61 bear=0.0 vAcc=??? sAcc=??? bAcc=??? mock]

**MockLocation:  Latitude in DB = 29.770000457763672
MockLocation:  Longitude in DB = 77.27999877929688**
`

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -6,6 +6,7 @@
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+    <uses-permission android:name="android.permission.ACCESS_MOCK_LOCATION" />
```
Comment (nikhil-roy) [inline app/src/main/AndroidManifest.xml]: Do we need this?

File: locationsyncservice/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,7 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.shuttl.locations_sync" >
+    package="com.shuttl.locations_sync">
 
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_MOCK_LOCATION"/>
```
Comment (nikhil-roy) [inline locationsyncservice/src/main/AndroidManifest.xml]: Do we need this?

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        )
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun subscribeMockLocationTracking(
+        mLocationManager: LocationManager,
+        provider: String,
+        locationListener: LocationListener
+    ) {
+        mockLocationListener = locationListener
```
Comment (nikhil-roy) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt]: why mockLocationListener = locationListener

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        mLocationManager.requestLocationUpdates(provider, 100, 0f, mockLocationListener)
```
Comment (nikhil-roy) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt]: should 100, 0f come from config?

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProviderManager.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    val PREF_KEY_MOCK_LOCATION_LNG = "mock_location_lng"
+
+    private fun getSharedPreferences(context: Context): SharedPreferences {
+        return context.getSharedPreferences(PREF_NAME_MOCK_LOCATION_PROVIDER, Context.MODE_PRIVATE)
+    }
+
+    fun getMockLocationProviderLatitude(context: Context?): Float {
+        requireNotNull(context)
+        val sharedPreferences = getSharedPreferences(context)
+        return sharedPreferences.getFloat(PREF_KEY_MOCK_LOCATION_LAT, 29.77.toFloat())
```
Comment (nikhil-roy) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProviderManager.kt]: need more context on this

File: locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt [hunk 6]
```diff
@@ -33,14 +33,17 @@ object LocationsHelper {
...
+        interceptor: Interceptor? = null,
+        locationConfigs: LocationConfigs
+    ) {
         setNetworkingDebug(interceptor)
         val pingIntent = Intent(app, LocationPingService::class.java)
         pingIntent.putExtra("config", locationConfigs)
         val saveIntent = Intent(app, LocationSaveService::class.java)
         saveIntent.putExtra("config", locationConfigs)
         app.startService(pingIntent)
-        app.startService(pingIntent)
```
Comment (nikhil-roy) [inline locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt]: add: app.startService(saveIntent)
remove: app.bindService(...

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (kshamashuttl) [reply app/src/main/AndroidManifest.xml]: Removed

File: locationsyncservice/src/main/AndroidManifest.xml (hunk 2 above)
Comment (kshamashuttl) [reply locationsyncservice/src/main/AndroidManifest.xml]: Removed

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt (hunk 3 above)
Comment (kshamashuttl) [reply locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt]: Removed

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt (hunk 4 above)
Comment (kshamashuttl) [reply locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProvider.kt]: Now these values are coming from config.

File: locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProviderManager.kt (hunk 5 above)
Comment (kshamashuttl) [reply locationsyncservice/src/main/java/com/shuttl/location_pings/mockLocation/MockLocationProviderManager.kt]: Added default value as 0.0

File: locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt (hunk 6 above)
Comment (kshamashuttl) [reply locationsyncservice/src/main/java/com/shuttl/location_pings/config/open_lib/LocationsHelper.kt]: Removed : bindService() and serviceConnection() method
Added : app.startService(saveIntent)

**Review by nikhil-roy — APPROVED** [review]

**Pull request merged**